core/types/calls.cc

```diff
@@ -995,12 +1002,42 @@ DispatchResult dispatchCallSymbol(const GlobalState &gs, const DispatchArgs &arg
                     continue;
                 }
                 consumed.insert(spec.name);
-                TypeAndOrigins tpe;
-                tpe.origins = {kwargsLoc};
+                // We want the precise location of the kwarg that we are processing.
+                // To do that, we rely on hash->{keys,values} being added in the order
+                // that the kwargs appear in the the call.  We also rely on args from
+                // a shapely-typed kwsplat hash (implicit or explicit) being added to
+                // hash->{keys,values} after the explicitly-named kwargs in the call.
                 auto offset = arg - hash->keys.begin();
+                // offset * 2 skips kwargs we have already seen.
+                // + 1 moves us to the value being passed.
+                ENFORCE(!sawKwSplat);
+                auto kwaValue = kwait + (offset * 2) + 1;
+                TypeAndOrigins tpe;
+                ENFORCE(args.args.size() == args.locs.args.size(), "what?!");
+                // This in-bounds check is unusual, but it's necessary for cases
+                // where we are processing a kwarg that comes from a non-kwsplat
+                // shapely-typed hash.  In such cases, our kwargOffset will appear to
+                // be out-of-bounds with respect to the actual number of arguments in
+                // the call.  (We could track the location information of kwargs from
+                // such a hash, but that would require rewriting a good chunk of this
+                // function.)  We actually do something slightly stronger, which is
+                // to check whether the thing we're pointing at lives inside the range
+                // of passed kwargs in the call.
+                //
+                // If we encounter this case, we've already setup kwargsLoc to reflect
+                // as closely as possible the location of the entire shapely-typed
+                // hash, so use that for the "origin" of the argument and say that the
+                // argument has no location.  The latter also inhibits type-driven
+                // autocorrects from kicking in.
+                bool argInBounds = kwaValue < ait;
+                // Note: it's OK to compute with kwaValue even if it's out-of-bounds.
```

Comment:
Because it's a random-access iterator?

Reply:
Sorry, I was being unclear here.  What this comment really intends to point out is that it's OK to compute a pointer (or an iterator, more generally) that's off the end of `args.args` and even to do pointer arithmetic on it to get an integer offset from the beginning of `args.args` -- so long as you don't actually dereference the memory at the pointer, or the memory that would live at `args.locs.args.begin() + kwargOffset` (which is effectively what `args.argLoc(kwargOffset)` is going to do in a line or two).  But expressing all of that is kind of a mouthful, so maybe we're better off dropping this note?

Reply:
Sure, that works for me. I think the reasoning about why it's safe makes sense 👍